cirq-core/cirq/transformers/dynamical_decoupling.py

```diff
@@ -149,34 +147,159 @@ def _merge_single_qubit_ops_to_phxz(
     return gate.on(q)
 
 
-def _calc_pulled_through(moment: Moment, input_pauli_ops: ops.PauliString) -> ops.PauliString:
-    """Calculates the pulled_through such that circuit(input_pauli_ops, moment.clifford_ops) is
-    equivalent to circuit(moment.clifford_ops, pulled_through).
+class _GateLabel(Enum):
+    UNKNOWN = 0
+    # Non-insertable gates that cannot be pulled through
+    WALL = 1
+    # Clifford gates where Pauli Gates can be pulled through
+    DOOR = 2
+    # An empty gate can be used to insert Pauli gates from the dd sequence
+    INSERTABLE = 3
+
+    def __str__(self) -> str:
+        match self:
+            case _GateLabel.UNKNOWN:
+                return '?'
+            case _GateLabel.WALL:
+                return 'w'
+            case _GateLabel.DOOR:
+                return 'd'
+            case _GateLabel.INSERTABLE:
+                return 'i'
+
+
+@frozen
+class _LabeledCircuit:
```

Comment:
I don't like the _LabeledCircuit/_LabeledGate names ... lets call them it _Grid,_CellType since we are turning the circuit into a 2 grid where each cell has one of 4 types.